app/views.py

```diff
@@ -201,3 +274,35 @@ def default(self, thing):
 
 def jsonize(x):
     return json.dumps(x, cls=DBModelJSONEncoder)
+
+# ----------------------------------------------------------- utility fns --- #
+
+def mkdirp(p):
+    if os.path.exists(p) and os.path.isdir(p):
+        return False
+    else:
+        os.makedirs(p)
+        return True
+
+
+def upload_dir(p):
+    if isinstance(p, basestring):
+        add_path = (p,) # Turn the single element into a tuple that can be
+                        # joined into a full path
+    elif isinstance(p, (list, tuple)):
+        add_path = p
+    else:
+        msg = 'I Don\'t know how to create an upload directory path from "%s", %s' \
+              % (p, type(p))
+        raise Exception(msg)
+    full_path = (app.config['UPLOAD_PATH'],) + add_path
+    return os.path.join(*full_path)
+
+
+def hashfile(fn, hasher, blocksize=65536):
```

Comment:
Is there any way for the client side to hash the individual parts, so we improve the performance for large files? Something to think about, but not something to implement near term.